### PR TITLE
Fix time indexer encoder and parser in fitting standardized index and…

### LINKED ISSUE #1842: Fitted paramers reloaded from NC lead to errors in standardized index, and setting multiple indexers leads to inconsistent behaviour
### Setup Information

- Xclim version: 0.51.0
- Python version: 3.11.8
- Operating System: Ubuntu 20.04

### Description

The time indexer in the `xc.indices.stats.standardized_index_fit_params` was not correctly stored in the attributes, leading to errors in different cases:
1. When there is no indexer, the indexer in the `xc.indices.stats.standardized_index_fit_params` was changed after `to_netcdf` and reloaded from `xr.open_dataset()`, leading to an error when running `standardized_precipitation_index(pr, params=params)`;
2. When the value of indexer is a non-array single value, there will be an error when encoding the attribute `params.attrs["time_indexer"]` in `stats.standardized_index_fit_params`;
3. When there are multiple indexers, only one indexer is saved, leading to inconsistent behaviour when running `standardized_precipitation_index(pr, params=params)`.

### Steps To Reproduce

Load dataset:
```
import xarray as xr
import xclim.indices as xci
from xclim.testing import open_dataset

ds = open_dataset("sdba/CanESM2_1950-2100.nc").isel(location=1)
pr = ds.pr.sel(time=slice("1998", "2000"))
pr_cal = ds.pr.sel(time=slice("1950", "1980"))
params = xci.stats.standardized_index_fit_params(
    pr_cal, freq="MS", window=1, dist="gamma", method="APP", fitkwargs={"floc": 0}, zero_inflated=True)
```
For the first issue,
```
params.to_dataset(name="params").to_netcdf("params.nc")
params = xr.open_dataset("params.nc").params

spi = xci.standardized_precipitation_index(pr, params=params)  # this line leads to the error
```
For the second issue,
```
params = xci.stats.standardized_index_fit_params(
    pr_cal, freq="MS", window=1, dist="gamma", method="APP", fitkwargs={"floc": 0}, zero_inflated=True, month=2)
# executable only when an array is assigned to `month`
```
For the third issue,
```
params = xci.stats.standardized_index_fit_params(
    pr_cal, freq="MS", window=1, dist="gamma", method="APP", fitkwargs={"floc": 0}, zero_inflated=True, drop=True, month=[2, 3])  # `drop=True` is not saved in params.attrs["time_indexer"]`
xci.standardized_precipitation_index(pr, params=params)  # the months other than 2 & 3 were not dropped
```

### Additional context

The error logs are attached here.
For the first issue,
```
File ~/.conda/envs/syncfuture/lib/python3.11/site-packages/xclim/indices/stats.py:882, in standardized_index(da, freq, window, dist, method, zero_inflated, fitkwargs, cal_start, cal_end, params, **indexer)
    [880](~/.conda/envs/syncfuture/lib/python3.11/site-packages/xclim/indices/stats.py:880) # Unpack attrs to None and {} if needed
    [881](~/.conda/envs/syncfuture/lib/python3.11/site-packages/xclim/indices/stats.py:881) freq = None if freq == "" else freq
--> [882](~/.conda/envs/syncfuture/lib/python3.11/site-packages/xclim/indices/stats.py:882) indexer = {} if indexer[0] == "" else {indexer[0]: indexer[1:]}
    [883](~/.conda/envs/syncfuture/lib/python3.11/site-packages/xclim/indices/stats.py:883) if cal_start or cal_end:
    [884](~/.conda/envs/syncfuture/lib/python3.11/site-packages/xclim/indices/stats.py:884)     warnings.warn(
    [885](~/.conda/envs/syncfuture/lib/python3.11/site-packages/xclim/indices/stats.py:885)         "Expected either `cal_{start|end}` or `params`, got both. The `params` input overrides other inputs."
    [886](~/.conda/envs/syncfuture/lib/python3.11/site-packages/xclim/indices/stats.py:886)         "If `cal_start`, `cal_end`, `freq`, `window`, and/or `dist` were given as input, they will be ignored."
    [887](~/.conda/envs/syncfuture/lib/python3.11/site-packages/xclim/indices/stats.py:887)     )

IndexError: string index out of range
```
For the second issue,
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[3], [line 1](vscode-notebook-cell:?execution_count=3&line=1)
----> [1](vscode-notebook-cell:?execution_count=3&line=1) params = xci.stats.standardized_index_fit_params(
      [2](vscode-notebook-cell:?execution_count=3&line=2)     pr_cal, freq="MS", window=1, dist="gamma", method="APP", fitkwargs={"floc": 0}, zero_inflated=True, month=2)

File ~/.conda/envs/syncfuture/lib/python3.11/site-packages/xclim/indices/stats.py:801, in standardized_index_fit_params(da, freq, window, dist, method, zero_inflated, fitkwargs, offset, **indexer)
    [791](~/.conda/envs/syncfuture/lib/python3.11/site-packages/xclim/indices/stats.py:791) params.attrs = {
    [792](~/.conda/envs/syncfuture/lib/python3.11/site-packages/xclim/indices/stats.py:792)     "calibration_period": cal_range,
    [793](~/.conda/envs/syncfuture/lib/python3.11/site-packages/xclim/indices/stats.py:793)     "freq": freq or "",
   (...)
    [798](~/.conda/envs/syncfuture/lib/python3.11/site-packages/xclim/indices/stats.py:798)     "units": "",
    [799](~/.conda/envs/syncfuture/lib/python3.11/site-packages/xclim/indices/stats.py:799) }
    [800](~/.conda/envs/syncfuture/lib/python3.11/site-packages/xclim/indices/stats.py:800) method, args = ("", []) if indexer == {} else indexer.popitem()
--> [801](~/.conda/envs/syncfuture/lib/python3.11/site-packages/xclim/indices/stats.py:801) params.attrs["time_indexer"] = (method, *args)
    [802](~/.conda/envs/syncfuture/lib/python3.11/site-packages/xclim/indices/stats.py:802) if offset:
    [803](~/.conda/envs/syncfuture/lib/python3.11/site-packages/xclim/indices/stats.py:803)     params.attrs["offset"] = offset
```

For the third issue, the attributes in the `params` are like 

and the obtained SPI would be like 
WX20240718-125333@2x.

### Contribution

- [X] I would be willing/able to open a Pull Request to address this bug.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by Hem-W**
### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #1842 
- [x] Tests for the changes have been added (for bug fixes / features)
  - [x] (If applicable) Documentation has been added / updated (for bug fixes / features)
- [x] CHANGELOG.rst has been updated (with summary of main changes)
  - [x] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added

### What kind of change does this PR introduce?

Fix the encoder and parser for the time indexer in the `xc.indices.stats.standardized_index_fit_params` and `xc.indices.stats.standardized_index` to allow no indexer, non-array indexers, and multiple indexers.

### Does this PR introduce a breaking change?

No. But the new parser is not compatible with the previous as the format of params.attrs["time_indexer"] is different. If any saved `params` from  previous version were reloaded, it may lead to errors.

### CODE REVIEW

**Review by aulemahal — APPROVED** [review]
This looks excellent!  Thanks a lot.

File: xclim/indices/stats.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 from collections.abc import Sequence
 from typing import Any
 
+import jsonpickle
```
Comment (aulemahal) [inline xclim/indices/stats.py:9]:
```suggestion
import json
```

File: xclim/indices/stats.py [hunk 2]
```diff
@@ -797,8 +798,7 @@ def standardized_index_fit_params(
         "group": group,
         "units": "",
     }
-    method, args = ("", []) if indexer == {} else indexer.popitem()
-    params.attrs["time_indexer"] = (method, *args)
+    params.attrs["time_indexer"] = jsonpickle.encode(indexer)  # noqa: S301
```
Comment (aulemahal) [inline xclim/indices/stats.py:801]:
```suggestion
    params.attrs["time_indexer"] = json.dumps(indexer)  # noqa: S301
```

File: xclim/indices/stats.py [hunk 3]
```diff
@@ -879,7 +879,7 @@ def standardized_index(
         )
         # Unpack attrs to None and {} if needed
         freq = None if freq == "" else freq
-        indexer = {} if indexer[0] == "" else {indexer[0]: indexer[1:]}
+        indexer = jsonpickle.decode(indexer)  # noqa: S301
```
Comment (aulemahal) [inline xclim/indices/stats.py:882]:
```suggestion
        indexer = json.loads(indexer)  # noqa: S301
```

**Review by aulemahal — COMMENTED** [review]
Now that I think of it, was there a need for using `jsonpickle` instead of the builtin `json` ? 

`jsonpickle` is a dependency in xclim because of some stuff in `sdba` where `json` can't do the job. When the plan of taking `sdba` out of xclim is executed, we'll want to remove unneeded dependencies.

**Comment by Hem-W** [comment]
> Now that I think of it, was there a need for using `jsonpickle` instead of the builtin `json` ?
> 
> `jsonpickle` is a dependency in xclim because of some stuff in `sdba` where `json` can't do the job. When the plan of taking `sdba` out of xclim is executed, we'll want to remove unneeded dependencies.

Thank you for pointing out. I believe `json` can meet the needs.

**Review by aulemahal — APPROVED** [review]
Approved again, thanks!

**Pull request merged**